File: Data Structures & Algorithms/count-prefix-and-suffix-pairs-i/submission-0.py
class Solution:
    def _isPandS(self, s1, s2):
        n1, n2 = len(s1), len(s2)
        ans = (s1 == s2[:n1]) and (s1 == s2[n2-n1:])
        return ans
    
    def countPrefixSuffixPairs(self, words: List[str]) -> int:
        n = len(words)
        count = 0
        for i in range(n):
            for j in range(i+1, n):
                if self._isPandS(words[i], words[j]):
                    count+= 1
        return count

File: Data Structures & Algorithms/word-break-ii/submission-0.cpp
class Solution {
    
    vector<string> ans;

public:
    void _backtracking(string& s, int i, unordered_set<string>& wordDict, vector<string>& res) {
        if (i == s.size()) {
            ans.push_back(join(res));
            return ; //something
        }

        for (int j = i; j < s.size(); ++j) {
            string substring = s.substr(i, j-i+1);
            if (wordDict.contains(substring)) {
                res.push_back(substring);
                _backtracking(s, j+1, wordDict, res);
                res.pop_back();
            }
        }
    }

    string join(const vector<string>& words) {
        ostringstream oss;
        for (int i = 0; i < words.size(); ++i) {
            if (i > 0) oss << " ";
            oss << words[i];
        }
        return oss.str();
    }
    
    vector<string> wordBreak(string s, vector<string>& wordDict) {
        unordered_set<string> wordSet(wordDict.begin(), wordDict.end());
        vector<string> res = {};
        _backtracking(s, 0, wordSet, res);
        return ans;
    }
};